keps/sig-architecture/3138-increase-reliability-bar/README.md

```diff
@@ -0,0 +1,415 @@
+clusters before.
+
+To make Kubernetes (or any other project) sustainable, we should ensure that
+existing architecture, individual features, tests etc. really works as expected.
+If there is work needed to improve reliability (or stability) of those, it
+should come before new features.
+This is often not the case currently. We significantly rely on the heroic
+efforts of very few individuals that are keeping us on the surface with new bugs,
+issues, test flakes etc. This approach isn’t sustainable and doesn’t scale.
+We need to make it a business of everyone in the whole community.
+
+[Production Readines Survey]: https://datastudio.google.com/reporting/2e9c7439-202b-48a9-8c57-4459e0d69c8d/page/GC5HB
+[How to Break your Cluster with networking]: https://www.youtube.com/watch?v=7qUDyQQ52Uk)
+[1]: https://www.youtube.com/watch?v=QKI-JRs2RIE
+[2]: https://www.youtube.com/watch?v=6sDTB4eV4F8
+[this example]: https://github.com/kubernetes/kubernetes/pull/98305
+
+
+### Goals
+
+- Define the high-level areas of investments
+- Define the sustainable process to ensure reliability investments
+- Define the details of the first milestone as a pilot (reducing test flakiness)
+
+
+### Non-Goals
+
+- Providing exact list of long-standing reliability issues (milestone 3) or new
+  reliability investments (milestone 4) - those will be handled by
+  Reliability WG separately
+
+
+## Proposal
+
+This document outlines steps to improve the quality and reliability of Kubernetes.
+Given the maturity level the project is reaching and the non-negligible amount of
+technical debt accumulated over the years, the most important aspect of the
+proposal is to build the culture which prioritizes certain types of contributions
+over purely feature work.
+
+We have to acknowledge that fixing, and especially debugging, reliability or
+stability issues is the kind of work that very few people like doing - for the
+majority it’s something that they would never do if they wouldn’t have to.
+This is especially true when it comes to debugging flaky tests.
+Moreover, while in companies there are multiple ways of enforcing that certain
+types of work is actually happening (e.g. oncall rotations, dedicated sprints, etc.),
+it’s much harder to achieve in the projects driven by the open-source community.
+
+As a result, we’re proposing introducing a process that will allow us to ensure
+that this kind of work that generally attracts fewer developers  will be really
+happening. This means, that for every piece of such work, we will introduce a two
+phase approach:
+1. announce the reliability-related issue to the SIG asking for prioritizing it
+   with a realistic (and non-conservative) deadline.
+   *We really hope that this will be effective in the majority of cases and
+   we will never be forced to use the point (2) special powers.*
+1. if the issue (reported above) is not addressed before the deadline, block
+   new features from the owning SIG until the issue will get addressed
```

Comment:
This will not work. This just moves the problem from one group which cannot solve it (the overall project) to another group which also cannot solve it (the SIG). SIGs don't have the capability to force people to do things, either!

Reply:
It seems like there are a few steps here:

1. Identifying difficult bugs and underwater SIGs.
2. Root-causing the difficult bugs.
3. Designing fixes, which for some issues might require some restructuring.
4. Implementing the fixes.
5. Reviewing and landing the fixes.

Processes that add management boilerplate like root-cause analysis and regression labeling and failure-rate tracking help with 1.  But processes take effort and time, which reduces the time folks have to actually fix things.  Hopefully identification work help you be smarter about picking the most impactful issues to work on.  But also, maybe it lets you phrase a very clear cry for help, in the hope that some of the many folks who are making money on top of Kubernetes notice and say "oh, I have some money/devs who can maybe help with this issue...".  And then you get a lift with 2 and 4.  And hopefully offloading those allows the core reviewers and approvers focus on 3 and 5.  And maybe some of the folks working on 2 and 4 can get sucked in as reviewers and eventually approvers.

Or maybe nobody comes to help :shrug:, in which case there's still a nicely prioritized stability backlog that just piles up for whenever folks who have time can burn it down ;).  But if folks are excited/invested enough to add new features, presumably there is somebody in that pipeline who is also interested in some of the identified stability issues.